readthedocs/redirects/models.py

```diff
@@ -67,3 +78,97 @@ def __unicode__(self):
                 self.to_url))
         else:
             return ugettext('Redirect: %s' % self.get_redirect_type_display())
+
+    def get_full_path(self, filename, language=None, version_slug=None):
+        """
+        Return a full path for a given filename. This will include version and
+        language information. No protocol/domain is returned.
+        """
+
+        # Handle explicit http redirects
+        if re.match('^https?://', filename):
+            return filename
+
+        url_kwargs = {
+            'project_slug': self.project.slug,
+            'filename': filename,
+        }
+
+        if not self.project.single_version:
+            if language is None:
+                language = self.project.language
+            if version_slug is None:
+                version_slug = self.project.get_default_version()
+            else:
+                versions = self.project.versions.all()
+                if not versions.filter(slug=version_slug).exists():
+                    version_slug = self.project.get_default_version()
+            url_kwargs.update({
+                'lang_slug': language,
+                'version_slug': version_slug,
+            })
```

Comment:
I feel like we keep rewriting this bit of code throughout the codebase, and should standardize on one approach and centralize it. It's in at least 3 places (core/views.py, project model get_docs_url, and redirects). I have started cleaning up the doc serving & subdomain_urls to standardize them in the subproject branch, and I think we should be able to pull this into one place.